src/python/pants/backend/jvm/tasks/coverage/jacoco.py

```diff
@@ -61,13 +100,69 @@ def classpath_prepend(self):
 
   @property
   def extra_jvm_options(self):
-    # TODO(jtrobec): implement code coverage using jacoco
-    return []
+    agent_option = '-javaagent:{agent}=destfile={destfile}'.format(agent=self._agent_path,
+                                                                   destfile=self._coverage_datafile)
+    return [agent_option]
+
+  @staticmethod
+  def is_coverage_target(tgt):
+    return (tgt.is_java or tgt.is_scala) and not tgt.is_test and not tgt.is_synthetic
 
   def report(self, execution_failed_exception=None):
-    # TODO(jtrobec): implement code coverage using jacoco
-    pass
+    if execution_failed_exception:
+      self._context.log.warn('Test failed: {0}'.format(execution_failed_exception))
+      if self._coverage_force:
+        self._context.log.warn('Generating report even though tests failed.')
+      else:
+        return
+
+    report_dir = os.path.join(self._settings.coverage_dir, 'reports')
+    safe_mkdir(report_dir, clean=True)
+    for report_format in ['xml', 'csv', 'html']:
+      target_path = os.path.join(report_dir, report_format)
+      args = ['report', self._coverage_datafile] + self.get_target_classpaths() + self.get_source_roots() + [
+        '--{report_format}={target_path}'.format(report_format=report_format,
+                                                 target_path=target_path)
+      ]
+      main = 'net.sourceforge.cobertura.reporting.ReportMain'
+      result = self._execute_java(classpath=self._cli_path,
+                                  main='org.jacoco.cli.internal.Main',
+                                  jvm_options=self._settings.coverage_jvm_options,
+                                  args=args,
+                                  workunit_factory=self._context.new_workunit,
+                                  workunit_name='jacoco-report-' + report_format)
+      if result != 0:
+        raise TaskError("java {0} ... exited non-zero ({1})"
+                        " 'failed to report'".format(main, result))
+
+  def get_target_classpaths(self):
```

Comment:
Good idea to prefix private properties/methods with `_`.

Reply:
done